decode.go

```diff
@@ -840,12 +840,13 @@ func (d Decoder) decodeStructFromTypeWith(typ Type, to reflect.Value, s *structT
 		if _, b, err = d.decodeTypeAndString(); err != nil {
 			return
 		}
+		k := string(b)
```

Comment:
There's a high chance that Go will make a dynamic memory allocation for this statement, can we modify by moving `f := s.fieldsByName[string(b)]` here?